src/ambianic/pipeline/ai/fall_detect.py

```diff
@@ -0,0 +1,110 @@
+"""Fall detection pipe element."""
+import logging
+from ambianic.pipeline.ai.image_detection import TFImageDetection
+from ambianic.pipeline.ai.pose_engine import PoseEngine
+from ambianic.util import stacktrace
+import math
+log = logging.getLogger(__name__)
+
+
+class FallDetector(TFImageDetection):
+    """Detects falls comparing two images spaced about 1-2 seconds apart."""
+    def __init__(self,
+                 model=None,
+                 **kwargs
+                 ):
+        """Initialize detector with config parameters.
+        :Parameters:
+        ----------
+        model: dict
+        {
+            'tflite': 
+                'ai_models/posenet_mobilenet_v1_075_721_1281_quant_decoder.tflite'
+            'edgetpu': 
+                'ai_models/posenet_mobilenet_v1_075_721_1281_quant_decoder_edgetpu.tflite'
+            labels: ai_models/pose_labels.txt
+        }
+        """
+        super().__init__(model, **kwargs)
+        self._prev_vals = []
+        self._pose_engine = PoseEngine(model, **kwargs)
+        self._fall_factor = 20
+
+    def process_sample(self, **sample):
+        """Detect objects in sample image."""
+        log.debug("%s received new sample", self.__class__.__name__)
+        if not sample:
+            # pass through empty samples to next element
+            yield None
+        else:
+            try:
+                image = sample['image']
+                inference_result = self.fall_detect(image=image)
+                inf_meta = {
+                    'display': 'Fall Detection',
+                }
+                # pass on the results to the next connected pipe element
+                processed_sample = {
+                    'image': image,
+                    'inference_result': inference_result,
+                    'inference_meta': inf_meta
+                    }
+                yield processed_sample
+            except Exception as e:
+                log.error('Error "%s" while processing sample. '
+                          'Dropping sample: %s',
+                          str(e),
+                          str(sample)
+                          )
+                log.warning(stacktrace())
+
+    def calculate_angle(self, p):
+        x1, y1 = p[0][0][0], p[0][0][1]
+        x2, y2 = p[0][1][0], p[0][1][1]
+        x3, y3 = p[1][0][0], p[1][0][1]
+        x4, y4 = p[1][1][0], p[1][1][1]
+        
+        m1 = (y1-y2)/(x1-x2)
+        m2 = (y3-y4)/(x3-x4)
+        
+        theta1 = math.degrees(math.atan(m1))
```

Comment:
@bhavikapanara do you take into account situations when the angle may be a negative number but still meets our criteria? 

Also, thinking some more about it, do we want the angle in the first image to be close to 90' (+/- 15') (implying the person is standing up) and then the angle in the second image to be more than +/- 60' relative to the first one? The idea here is to prevent situations where a person is lying down and they stand up, which will be a more than 60' change of angle but they would be just getting off a couch or a bed, instead of falling down or laying down.

I don't think we need to go too far with these empirical rules, because it will turn into a spaghetti code. Instead we can start a new project (after this PR) to design an RNN that predicts fall based on a sequence of poses.

Your thoughts will be much appreciated.

Reply:
Thank you for the update @bhavikapanara . Closing this discussion.